File: ProblemSolving/ProblemSolving/PG42579_베스트앨범.cpp
// 시간복잡도 : O(N*M*logN). N은 genres배열 갯수, M은 장르 종류 수

#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

struct P {
    int idx, cnt;
};

struct N {
    string genre;
    int cnt;
};

bool cmpP(P a, P b) {
    return a.cnt > b.cnt;
}

bool cmpN(N a, N b) {
    return a.cnt > b.cnt;
}

unordered_map<string, vector<P>> um;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;

    // 해시테이블에 장르별 재생횟수 데이터 저장
    // Hash[key] = [ {-1, 총 재생횟수}, {idx, cnt}, {idx, cnt}, ...]
    for (int i = 0; i < genres.size(); i++) {           // O(N)
        // 새로운 장르이면
        if (um.find(genres[i]) == um.end())
            um[genres[i]].push_back({ -1, 0 });
        
        // 장르명을 Key로 id, 재생횟수를 Value로 삽입
        um[genres[i]].push_back({ i, plays[i] });
        um[genres[i]][0].cnt += plays[i];   // 재생횟수 가산
    }

    // 장르별 총 재생횟수값 추출
    vector<N> v;
    unordered_map<string, vector<P>>::iterator it;
    for (it = um.begin(); it != um.end(); it++) {      // O(M)
        // 장르별 총 재생횟수를 vector에 삽입
        v.push_back({ it->first, (it->second)[0].cnt });
        
        // 같은 장르에서는 재생 횟수를 기준으로 내림차순 정렬
        sort(um[it->first].begin(), um[it->first].end(), cmpP);     // O(N * logN)
    }

    // 내림차순 정렬
    sort(v.begin(), v.end(), cmpN);

    // 한 장르 내에서 가장 많이 재생한 노래 idx 추출
    for (int i = 0; i < v.size(); i++) {            // O(M)
        int num = um[v[i].genre].size();    // 해당 장르의 노래 수
        if (num > 3) num = 3;

        for (int j = 1; j < num; j++)               // O(2)
            answer.push_back(um[v[i].genre][j].idx);
    }

    return answer;
}

int main() {

    vector<string> genres = { "classic", "pop", "classic", "classic", "pop" };
    vector<int> plays = { 500, 600, 150, 800, 2500 };
    
    vector<int> answer = solution(genres, plays);
    for (int& a : answer)
        cout << a << " ";

}

File: ProblemSolving_Java/ProblemSolving_Java/src/programmers/level3/PG42579_베스트앨범.java
package programmers.level3;

import java.util.*;

public class PG42579_베스트앨범 {

    public static void main(String[] args) {

        String[] genres = {"classic", "pop", "classic", "classic", "pop"};
        int[] plays = {500, 600, 150, 800, 2500};

        Solution solution = new Solution();
        int[] solution1 = solution.solution(genres, plays);
        for (int i : solution1) {
            System.out.println("i = " + i);
        }
    }

    static class Song {
        int idx;    // 몇 번째 노래가
        int cnt;    // 몇 번 재생되었는가

        Song(int idx, int cnt) {
            this.idx = idx;
            this.cnt = cnt;
        }
    }

    static class Solution {

        Map<String, Integer> playCntHashMap = new HashMap<>();      // 장르별 노래 재생 횟수
        Map<String, List<Song>> songListHashMap = new HashMap<>();  // 어느 장르에 속하는 노래 리스트

        public int[] solution(String[] genres, int[] plays) {
            List<Integer> ansList = new ArrayList<>();

            for(int i=0; i<genres.length; i++) {
                String name = genres[i];
                int cnt = plays[i];
                
                // 장르별 재생 횟수를 누적
                playCntHashMap.put(name, playCntHashMap.getOrDefault(name, 0)+cnt);

                // 장르별 노래 리스트에 추가
                List<Song> songList = songListHashMap.getOrDefault(name, new ArrayList<Song>());
                songList.add(new Song(i, cnt));
                songListHashMap.put(name, songList);
            }

            // 누적 재생 횟수별로 장르명 정렬하기
            List<Map.Entry<String, Integer>> list = new ArrayList<>(playCntHashMap.entrySet());
            Collections.sort(list, new Comparator<Map.Entry<String, Integer>>() {
                @Override
                public int compare(Map.Entry<String, Integer> o1, Map.Entry<String, Integer> o2) {
                    return o2.getValue()-o1.getValue();
                }
            });

            Iterator<Map.Entry<String, Integer>> iterator = list.iterator();
            while(iterator.hasNext()) {

                Map.Entry<String, Integer> next = iterator.next();

                List<Song> songs = songListHashMap.get(next.getKey());

                Collections.sort(songs, new Comparator<Song>() {
                    @Override
                    public int compare(Song o1, Song o2) {
                        if(o1.cnt == o2.cnt) {
                            return o1.idx - o2.idx;
                        }
                        return o2.cnt - o1.cnt;
                    }
                });

                ansList.add(songs.get(0).idx);
                if(songs.size() >= 2)
                    ansList.add(songs.get(1).idx);
            }

            Integer[] answer = new Integer[ansList.size()];
            ansList.toArray(answer);

            return Arrays.stream(answer).mapToInt(Integer::intValue).toArray();
        }
    }
}
